crates/bevy_math/src/cubic_splines.rs

```diff
@@ -472,22 +472,40 @@ impl<P: Point> CubicCurve<P> {
     /// A flexible iterator used to sample curves with arbitrary functions.
     ///
     /// This splits the curve into `subdivisions` of evenly spaced `t` values across the
-    /// length of the curve from start (t = 0) to end (t = 1), returning an iterator that evaluates
-    /// the curve with the supplied `sample_function` at each `t`.
+    /// length of the curve from start (t = 0) to end (t = n), where `n = self.segment_count()`,
+    /// returning an iterator evaluating the curve with the supplied `sample_function` at each `t`.
     ///
-    /// Given `subdivisions = 2`, this will split the curve into two lines, or three points, and
-    /// return an iterator over those three points, one at the start, middle, and end.
+    /// For `subdivisions = 2`, this will split the curve into two lines, or three points, and
+    /// return an iterator with 3 items, the three points, one at the start, middle, and end.
     #[inline]
-    pub fn iter_samples(
-        &self,
+    pub fn iter_samples<'a, 'b: 'a>(
+        &'b self,
         subdivisions: usize,
-        sample_function: fn(&Self, f32) -> P,
-    ) -> impl Iterator<Item = P> + '_ {
-        (0..=subdivisions).map(move |i| {
-            let segments = self.segments.len() as f32;
-            let t = i as f32 / subdivisions as f32 * segments;
-            sample_function(self, t)
-        })
+        mut sample_function: impl FnMut(&Self, f32) -> P + 'a,
+    ) -> impl Iterator<Item = P> + 'a {
+        self.iter_samples_t(subdivisions)
+            .map(move |t| sample_function(self, t))
+    }
+
+    /// Iterates over `subdivisions + 1` `t`s evenly spaced from the start of
+    /// the first curve segment to the end of the last segment.
+    #[inline]
+    fn iter_samples_t(&self, subdivisions: usize) -> impl ExactSizeIterator<Item = f32> {
+        let segments = self.segment_count() as f32;
+        let step = segments / subdivisions as f32;
+        (0..subdivisions.saturating_add(1)).map(move |i| i as f32 * step)
```

Comment:
```suggestion
        (0..=subdivisions).map(move |i| i as f32 * step)
```

I think this is equivalent?

Reply:
This won't add 1 to the start of the range. You would need to do 1..=subdivisions to be equivalent.

Reply:
I made it a `0..foo+1` because `0..=foo` doesn't implement `ExactSizeIterator`. The `ExactSizeIterator` might explain the drastic improvement in the `build_*_cubic_100_points` benchmarks. I'll check and report back.

Reply:
We are good! The performance improvement sticks even with the inclusive range, I switched to using it.